src/main/java/com/opentable/metrics/jvm/FileDescriptorMetricSet.java

```diff
@@ -15,88 +15,51 @@
- * file descriptors from the OS.
- */
 public class FileDescriptorMetricSet implements MetricSet {
-    private static final Logger LOG = LoggerFactory.getLogger(FileDescriptorMetricSet.class);
-    private final List<String> FDTYPE = ImmutableList.of("open", "max");
-    private final List<String> EXCEPTION = ImmutableList.of("illegal-access", "invocation-target");
-    private final OperatingSystemMXBean os = ManagementFactory.getOperatingSystemMXBean();
-    private final Map<String, Map<String, Counter>> calls;
-    private final Method openMethod;
-    private final Method maxMethod;
+    private static final Gauge<Long> NULL = () -> null;
 
-    public FileDescriptorMetricSet() {
-        final Supplier<Map<String, Counter>> makeCounters = () ->
-                ImmutableMap.of(
-                        "illegal-access", new Counter(),
-                        "invocation-target", new Counter()
-                );
-        calls = ImmutableMap.of(
-                "open", makeCounters.get(),
-                "max", makeCounters.get()
-        );
+    private final Map<String, Metric> metricMap;
 
-        openMethod = retrieveMethod("getOpenFileDescriptorCount");
-        maxMethod = retrieveMethod("getMaxFileDescriptorCount");
+    public FileDescriptorMetricSet() {
+        final OperatingSystemMXBean osBean = ManagementFactory.getOperatingSystemMXBean();
+        if (osBean instanceof UnixOperatingSystemMXBean) {
+            final UnixOperatingSystemMXBean bean = (UnixOperatingSystemMXBean) osBean;
+            metricMap = ImmutableMap.of(
+                    "open", caller(bean::getOpenFileDescriptorCount),
+                    "max", caller(bean::getMaxFileDescriptorCount)
+            );
+        } else {
+            metricMap = ImmutableMap.of(
+                    "open", NULL,
+                    "max", NULL
+            );
+        }
     }
 
     @Override
     public Map<String, Metric> getMetrics() {
-        final Map<String, Metric> gauges = new HashMap<>();
-
-        gauges.put(FDTYPE.get(0), (Gauge<Long>)() -> invoke(openMethod, FDTYPE.get(0)));
-        gauges.put(FDTYPE.get(1), (Gauge<Long>)() -> invoke(maxMethod, FDTYPE.get(1)));
-
-        for (String type: FDTYPE) {
-            for(String ex: EXCEPTION) {
-                gauges.put(String.format("%s.%s.%s", "exception", type, ex), calls.get(type).get(ex));
-            }
-        }
-        return Collections.unmodifiableMap(gauges);
+        return metricMap;
     }
 
-    private Method retrieveMethod(String name) {
-        try {
-            final Method method = os.getClass().getDeclaredMethod(name);
-            method.setAccessible(true);
-            return method;
-        } catch (NoSuchMethodException e) {
-            LOG.error("No such method exists:", e);
-            return null;
-        }
+    /**
+     * It's unclear whether or not {@link com.sun.management.UnixOperatingSystemMXBean} methods are re-entrant,
```

Comment:
https://docs.oracle.com/javase/7/docs/api/java/lang/management/package-summary.html

>  The java.lang.management API is thread-safe.

Reply:
Nice.  Simplified in 9df60ceb5e19a4dac9bf8feb96b02ecbff88cf7e.